File: AstrolabeProtocolKotlin/src/main/kotlin/dev/astrolabe/protocol/inspection/RuntimeLayoutRelation.kt
//
//  RuntimeLayoutRelation.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/20.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Node-owned normalized layout anchor. */
@Serializable
public data class RuntimeLayoutAnchor(
    /** Node owning this layout anchor. */
    public val nodeID: RuntimeOpaqueIdentifier,
    /** Open normalized anchor name. */
    public val anchor: String
) {
    init {
        require(anchor.isNotEmpty() && anchor.length <= MAXIMUM_LAYOUT_ANCHOR_LENGTH) {
            "Layout anchor must contain between 1 and 128 characters"
        }
    }
}

/** Closed comparison relation used by a layout constraint. */
@Serializable
public enum class RuntimeLayoutRelationKind {
    lessThanOrEqual,
    equal,
    greaterThanOrEqual
}

/** Platform-neutral layout relation. */
@Serializable
public data class RuntimeLayoutRelation(
    /** Producer-defined relation identifier when available. */
    public val identifier: String?,
    /** Source node anchor constrained by the relation. */
    public val source: RuntimeLayoutAnchor,
    /** Closed comparison relation. */
    public val relation: RuntimeLayoutRelationKind,
    /** Target node anchor, or null for a constant relation. */
    public val target: RuntimeLayoutAnchor?,
    /** Target coefficient applied before offset. */
    public val multiplier: Double,
    /** Logical-unit constant offset. */
    public val offset: RuntimeMeasurement,
    /** Normalized relation strength when available. */
    public val strength: Double?,
    /** Active state when reported by the platform. */
    public val active: Boolean?,
    /** Namespaced platform-specific relation facts. */
    public val extensions: RuntimeExtensionMap
) {
    init {
        require(strength == null || strength in 0.0..1.0) {
            "Layout relation strength must be between zero and one"
        }
        require(offset.unit == RuntimeMeasurementUnit.logical) {
            "Layout relation offset must use logical units"
        }
    }
}

private const val MAXIMUM_LAYOUT_ANCHOR_LENGTH: Int = 128

File: Implementations/Swift/Sources/AstrolabeProtocol/Inspection/RuntimeLayoutRelation.swift
//
//  RuntimeLayoutRelation.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/7/16.
//

public struct RuntimeLayoutAnchor: Codable, Equatable, Sendable {
    /// Node owning this layout anchor.
    public let nodeID: RuntimeOpaqueIdentifier

    /// Open normalized anchor name.
    public let anchor: String

    public init(nodeID: RuntimeOpaqueIdentifier, anchor: String) {
        self.nodeID = nodeID
        self.anchor = anchor
    }
}

public enum RuntimeLayoutRelationKind: String, Codable, Sendable {
    case lessThanOrEqual
    case equal
    case greaterThanOrEqual
}

public struct RuntimeLayoutRelation: Codable, Equatable, Sendable {
    /// Producer-defined relation identifier when available.
    @RuntimeRequiredNullable public private(set) var identifier: String?

    /// Source node anchor constrained by the relation.
    public let source: RuntimeLayoutAnchor

    /// Closed comparison relation.
    public let relation: RuntimeLayoutRelationKind

    /// Target node anchor, or nil for a constant relation.
    @RuntimeRequiredNullable public private(set) var target: RuntimeLayoutAnchor?

    /// Target coefficient applied before offset.
    public let multiplier: Double

    /// Logical-unit constant offset.
    public let offset: RuntimeMeasurement

    /// Normalized relation strength from zero to one when available.
    @RuntimeRequiredNullable public private(set) var strength: Double?

    /// Active state when reported by the platform.
    @RuntimeRequiredNullable public private(set) var active: Bool?

    /// Namespaced platform-specific relation facts.
    public let extensions: RuntimeExtensionMap

    public init(
        identifier: String?,
        source: RuntimeLayoutAnchor,
        relation: RuntimeLayoutRelationKind,
        target: RuntimeLayoutAnchor?,
        multiplier: Double,
        offset: RuntimeMeasurement,
        strength: Double?,
        active: Bool?,
        extensions: RuntimeExtensionMap
    ) {
        self.identifier = identifier
        self.source = source
        self.relation = relation
        self.target = target
        self.multiplier = multiplier
        self.offset = offset
        self.strength = strength
        self.active = active
        self.extensions = extensions
    }
}
